resharper/src/resharper-unity/CSharp/Daemon/Stages/Analysis/PreferGenericMethodOverloadAnalyzer.cs

```diff
@@ -0,0 +1,191 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using JetBrains.Annotations;
+using JetBrains.Application.UI.Utils;
+using JetBrains.DocumentModel;
+using JetBrains.Metadata.Reader.API;
+using JetBrains.ReSharper.Daemon.BuildScripts.DaemonStage.Highlightings;
+using JetBrains.ReSharper.Feature.Services.Daemon;
+using JetBrains.ReSharper.Feature.Services.Navigation.Goto.Filters;
+using JetBrains.ReSharper.Intentions.CSharp.QuickFixes;
+using JetBrains.ReSharper.Intentions.QuickFixes;
+using JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Errors;
+using JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Dispatcher;
+using JetBrains.ReSharper.Psi;
+using JetBrains.ReSharper.Psi.CSharp;
+using JetBrains.ReSharper.Psi.CSharp.Impl.DocComments;
+using JetBrains.ReSharper.Psi.CSharp.Parsing;
+using JetBrains.ReSharper.Psi.CSharp.Tree;
+using JetBrains.ReSharper.Psi.CSharp.Util;
+using JetBrains.ReSharper.Psi.CSharp.Util.Literals;
+using JetBrains.ReSharper.Psi.Impl.CodeStyle;
+using JetBrains.ReSharper.Psi.Resolve;
+using JetBrains.ReSharper.Psi.Resolve.Managed;
+using JetBrains.ReSharper.Psi.Tree;
+using JetBrains.ReSharper.Psi.Util;
+using JetBrains.Util;
+using JetBrains.Util.Extension;
+
+namespace JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Analysis
+{
+    [ElementProblemAnalyzer(typeof(IInvocationExpression), HighlightingTypes =
+        new[] {typeof(PreferGenericMethodOverloadWarning), typeof(AmbiguousTypeInStringLiteralWarning),
+            typeof(UnknownTypeInStringLiteralWarning)})] 
+    public class PreferGenericMethodOverloadAnalyzer : UnityElementProblemAnalyzer<IInvocationExpression>
+    {
+        private delegate bool TypeFilter(ITypeElement typeElement);
+        private static readonly IDictionary<string, TypeFilter> ourInterestingMethods = new Dictionary<string, TypeFilter>()
+        {
+            {"GetComponent", GetComponentTypeFilter},
+            {"AddComponent", AddComponentTypeFilter},
+            {"CreateInstance", CreateInstanceTypeFilter}
+        };
+        
+        private static readonly ISet<IClrTypeName> ourInterestingClasses = new HashSet<IClrTypeName>()
+        {
+            KnownTypes.ScriptableObject,
+            KnownTypes.GameObject,
+            KnownTypes.Component,
+        };
+        
+        public PreferGenericMethodOverloadAnalyzer(UnityApi unityApi)
+            : base(unityApi)
+        {
+        }
+
+        protected override void Analyze(IInvocationExpression expression, ElementProblemAnalyzerData data,
+            IHighlightingConsumer consumer)
+        {
+            if (expression.RPar == null) return;
+            if (expression.ContainsPreprocessorDirectives()) return;
```

Comment:
1. This check is considerably more expensive than the subsequent ones that are likely to filter out 95%+ of invocations that shouldn't be processed. Consider moving it somewhere later in the method.

2. It probably should only be checked for `PreferGenericMethodOverloadWarning`. Other warnings e.g. `UnknownTypeInStringLiteralWarning` look valid regardless of presence of inactive preprocessor branches. This would also solve the first problem.